### PR TITLE
[SPARK-46743][SQL][FOLLOW UP] Count bug after ScalarSubqery is folded if it has an empty relation

### PR DESCRIPTION
**Pull request by andylam-db**
### What changes were proposed in this pull request?

In this PR https://github.com/apache/spark/pull/45125, we handled the case where an Aggregate is folded into a Project, causing a count bug. We missed cases where:
1. The entire ScalarSubquery's plan is regarded as empty relation, and is folded completely. 
2. There are operations above the Aggregate in the subquery (such as filter and project).

### Why are the changes needed?

This PR fixes that by adding the case handling in ConstantFolding and OptimizeSubqueries.

### Does this PR introduce _any_ user-facing change?

Yes. There was a correctness error which happens when the scalar subquery is count-bug-susceptible, and empty, and thus folded by `ConstantFolding`. 

### How was this patch tested?

Added SQL query tests in `scalar-subquery-count-bug.sql`.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by andylam-db** [comment]
Ping for reviews! @agubichev @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -89,6 +89,12 @@ object ConstantFolding extends Rule[LogicalPlan] {
           e
       }
 
+    // Don't replace ScalarSubquery if its plan is an aggregate that may suffer from a COUNT bug.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:92]: I'm not a big fan of these band-aid fixes but I don't have a better idea. We should think about how to handle the count bug with a more general approach.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/subquery/scalar-subquery/scalar-subquery-count-bug.sql [hunk 2]
```diff
@@ -82,6 +82,73 @@ SELECT
 FROM
   l;
 
+-- Count bug over a subquery with an empty relation after optimization.
```
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/subquery/scalar-subquery/scalar-subquery-count-bug.sql:85]: Thank you for adding a test case.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]

File: sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out [hunk 3]
```diff
@@ -205,6 +205,122 @@ struct<scalarsubquery(a):bigint>
...
+struct<scalarsubquery(a):bigint>
+-- !query output
+0
+0
+0
+0
+0
+0
+0
+0
```
[the remark this answers is not in the document]
Comment (andylam-db) [reply sql/core/src/test/resources/sql-tests/results/subquery/scalar-subquery/scalar-subquery-count-bug.sql.out:233]: Looks like it's failing when decorrelation is turned off. I'll look into it. The generated answer (with zeroes) is correct.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 1 above)
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala:92]: I think the count bug is just a symptom, the main problem is that we handle decorrelation in 2 different places, and have the OptimizeSubqueries between them.

**Review by agubichev — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -357,19 +360,28 @@ abstract class Optimizer(catalogManager: CatalogManager)
...
         }
+        val newPlan = Project(projections,
+          if (predicates.nonEmpty) Filter(predicates.reduce(And), optimizedAgg) else optimizedAgg
+        )
+        val needTopLevelProject = newPlan.output.zip(optimizedAgg.output).exists {
+          case (oldAttr, newAttr) => oldAttr.exprId != newAttr.exprId
+        }
+        s.withNewPlan(
+          if (needTopLevelProject) newPlan else newPlan.child
+        )
```
Comment (andylam-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:384]: Let me know if this is janky.

**Review by agubichev — APPROVED** [review]

**Comment by andylam-db** [comment]
Looks like the CI build failures are unrelated.

**Comment by cloud-fan** [comment]
The CI failure is unrelated, thanks, merging to master!

**Pull request closed without merging**